src/psc_datalogger/connection.py

```diff
@@ -0,0 +1,410 @@
+        self._worker.query_complete.connect(status_bar.query_complete_callback)
+        self._worker.init_complete.connect(status_bar.init_complete_callback)
+        self._worker.error.connect(status_bar.error_callback)
+
+    def set_interval(self, interval: str):
+        """Set the update interval for the logging thread"""
+        if interval != "":  # Empty string passed if the textbox is cleared
+            self._worker.interval = float(interval)
+
+    def set_filepath(self, filepath):
+        """Set the filepath for the logging thread"""
+        self._worker.set_filepath(filepath)
+
+    def set_instrument(
+        self, instrument_number: int, gpib_address: str, measure_temp: bool
+    ):
+        """Configure the given instrument number with the provided parameters"""
+        self._worker.set_instrument(instrument_number, gpib_address, measure_temp)
+
+    def start_logging(self) -> bool:
+        """Start the logging process in the background thread.
+        Returns True if logging successfully started, else False"""
+        if self._worker.validate_parameters():
+            self.logging_signal.set()
+            self.status_bar.logging_started()
+            return True
+        else:
+            return False
+
+    def stop_logging(self) -> None:
+        """Stop the logging process in the background thread"""
+        self.logging_signal.clear()
+        self.status_bar.logging_stopped()
+
+
+@dataclass
+class InstrumentConfig:
+    """Contains the configuration for a single instrument"""
+
+    # GPIB address of instrument
+    address: int = -1
+    # Indicate whether the voltage read should be converted into a temperature
+    measure_temp: bool = False
+
+
+class DataWriter(DictWriter):
+    """Class that handles writing data to a file"""
+
+    file: TextIO
+
+    csv_fieldnames = [
+        "timestamp",
+        "instrument 1",
+        "instrument 2",
+        "instrument 3",
+    ]
+
+    def __init__(self, filepath: str):
+        self.file = open(filepath, "w")
+
+        super().__init__(self.file, fieldnames=self.csv_fieldnames, dialect="excel")
+
+        self.writeheader()
+        self.file.flush()
+
+    def close(self) -> None:
+        self.file.close()
+
+    def write(self, timestamp: datetime, ins_1: str, ins_2: str, ins_3: str):
+        """Write the given data to the file"""
+        len_written = self.writerow(
+            {
+                self.csv_fieldnames[0]: str(timestamp),
+                self.csv_fieldnames[1]: ins_1,
+                self.csv_fieldnames[2]: ins_2,
+                self.csv_fieldnames[3]: ins_3,
+            }
+        )
+        logging.debug(f"DictWriter wrote {len_written} bytes")
+
+        self.file.flush()
+        logging.debug("File flushed")
+
+
+class PrologixNotFoundException(Exception):
+    """Exception thrown when the Prologix controller is not found"""
+
+
+class Worker(QObject):
+    """Class that does the serial connection to the instruments
```

Comment:
I don't think this follows PEP 257. You need a newline between the first line and others.

This is also the case elsewhere in your PR.

Reply:
Thanks, fixed in 74038c2f4d71773bc8abc16969b240ab6d2bd04f